code/binary-tree/ViewModel/src/main/java/ru/unn/agile/binarytree/viewmodel/ViewModel.java

```diff
@@ -0,0 +1,218 @@
+    private final StringProperty removeKey = new SimpleStringProperty();
+
+    private final StringProperty addStatus = new SimpleStringProperty();
+
+    private final StringProperty findResult = new SimpleStringProperty();
+    private final StringProperty findStatus = new SimpleStringProperty();
+
+    private final StringProperty removeResult = new SimpleStringProperty();
+    private final StringProperty removeStatus = new SimpleStringProperty();
+
+    public ViewModel() {
+        addKey.set("");
+        addValue.set("");
+        addStatus.set(Status.WAITING.toString());
+
+        findKey.set("");
+        findResult.set("");
+        findStatus.set(Status.WAITING.toString());
+
+        removeKey.set("");
+        removeResult.set("");
+        removeStatus.set(Status.WAITING.toString());
+
+        // Add listeners to the input text fields
+        final List<StringProperty> fields = new ArrayList<>() {
+            {
+                add(addKey);
+                add(addValue);
+                add(findKey);
+                add(removeKey);
+            }
+        };
+
+        for (StringProperty field : fields) {
+            final ValueChangeListener listener = new ValueChangeListener();
+            field.addListener(listener);
+            valueChangedListeners.add(listener);
+        }
+    }
+
+    public StringProperty addKeyProperty() {
+        return addKey;
+    }
+
+    public StringProperty addValueProperty() {
+        return addValue;
+    }
+
+    public StringProperty addStatusProperty() {
+        return addStatus;
+    }
+
+    public StringProperty findKeyProperty() {
+        return findKey;
+    }
+
+    public StringProperty findResultProperty() {
+        return findResult;
+    }
+
+    public StringProperty findStatusProperty() {
+        return findStatus;
+    }
+
+    public StringProperty removeKeyProperty() {
+        return removeKey;
+    }
+
+    public StringProperty removeResultProperty() {
+        return removeResult;
+    }
+
+    public StringProperty removeStatusProperty() {
+        return removeStatus;
+    }
+
+    public String getAddStatus() {
+        return addStatus.get();
+    }
+
+    public String getFindResult() {
+        return findResult.get();
+    }
+
+    public String getFindStatus() {
+        return findStatus.get();
+    }
+
+    public String getRemoveResult() {
+        return removeResult.get();
+    }
+
+    public String getRemoveStatus() {
+        return removeStatus.get();
+    }
+
+    public void add() {
+        final int key = Integer.parseInt(addKey.get());
+        final String value = addValue.get();
+
+        storage.add(key, value);
+
+        addStatus.set(Status.SUCCESS.toString());
+    }
+
+    public void find() {
+        final int key = Integer.parseInt(findKey.get());
+
+        final String result = storage.find(key);
+
+        findResult.set(result);
+        findStatus.set(Status.SUCCESS.toString());
```

Comment:
SUCCESS is a status of operation. Null - is a valid result of operation. We found nothing. 
There are several statuses available for the operation, independently from the value result:
WAITING, READY, BAD_FORMAT, SUCCESS - they are only for the tracking of the operation status

Reply:
Ok, but user will not understand that we don't found this value. 
1. Key exists - show "value" and print success
2. Key not exists - show "" and print success. And if user doesn't know about value field, he will not understand that key not exists. 
3. Key exists, value is "" - show "" and print success 

And as you can see, second and third are the same.

Reply:
Thanks for the showcase.
I've added dedicated message for the user. As we discussed, we cannot put it in the status, so that's a workaround which OK for lab work from my point of view